exporter/signalfxexporter/translation/hostdims.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package translation
+
+import (
+	"fmt"
+
+	sfxpb "github.com/signalfx/com_signalfx_metrics_protobuf/model"
+	"go.opentelemetry.io/collector/translator/conventions"
+)
+
+// getCloudDimension will return a dimension object if a cloud id (AWSUniqueId, gcp_id, etc.)
+// can be constructed from the provided metadata. If the values are not available return
+// value is nil.
+func getCloudDimension(resourceAttr map[string]string) *sfxpb.Dimension {
+	accountID := resourceAttr[conventions.AttributeCloudAccount]
+	region := resourceAttr[conventions.AttributeCloudRegion]
+	instanceID := resourceAttr[conventions.AttributeHostID]
+	provider := resourceAttr[conventions.AttributeCloudProvider]
+
+	var dimKey, dimValue string
+
+	switch provider {
+	// TODO: Should these be defined as constants in resourcedetector module that we import or somewhere else?
+	case "ec2":
+		if instanceID == "" || region == "" || accountID == "" {
+			return nil
+		}
+		dimKey = "AWSUniqueId"
+		dimValue = fmt.Sprintf("%s_%s_%s", instanceID, region, accountID)
+		delete(resourceAttr, conventions.AttributeCloudAccount)
+		delete(resourceAttr, conventions.AttributeCloudRegion)
+		delete(resourceAttr, conventions.AttributeHostID)
+	case "gce":
+		if accountID == "" || instanceID == "" {
+			return nil
+		}
+		dimKey = "gcp_id"
+		dimValue = fmt.Sprintf("%s_%s", accountID, instanceID)
+		delete(resourceAttr, conventions.AttributeCloudAccount)
+		delete(resourceAttr, conventions.AttributeHostID)
+	default:
+		return nil
+	}
+	delete(resourceAttr, conventions.AttributeCloudProvider)
```

Comment:
We should not mutate original resource object here. If user has another exporter in the pipeline, it'll get corrupted data

Reply:
Ok, made it copy the resource attribute map instead of reuse.

Reply:
What do you think of converting resource labels as is with `appendAttributesToDimensions` first and then get unique ids from dimensions? An extra map allocation seems to be more expensive than few of `sfxpb.Dimension` structs.

Reply:
I would be surprised if it made any significant difference even at high volume since there is one shared resource for a potentially large number of metrics/timeseries/datapoints.  If this were a hotspot I would be more concerned about it.

Reply:
Why do you think it must be large number of metrics? Host metrics usually doesn't have more than 10 metrics per resource. And during batching, we don't batch per resource, it's just one slice of metrics along with their resources. So I'd rather call it a hotspot.

Reply:
@jrcamp @bogdandrutu do you have any thoughts on this ^ ?

Reply:
And now we will have an extra map allocation for every resource, even if it won't be used to generate the unique id ...

Reply:
I'm just more going off my experience of having gone down the road of trying to optimize things when it seems like something obviously would help and not being able to find any statistically meaningful improvements and then being left with code that is significantly more difficult to understand for somebody who isn't familiar with it.  By the time you account for the extra allocations of dims, and the CPU usage of having to iterate over every dim and copy dims in the slice array to collapse the empty spots where you deleted a dim, I would be very surprised if there were real CPU savings but the code would definitely be more complicated.

Reply:
I'm sorry, I don't really want to argue on this that far, but it doesn't seem to be easy to understand already.
For example, looking at `cloudDim := getCloudDimension(resourceAttribs);` I would never expect that `resourceAttribs` would be mutated without looking into the function implementation. So the function doesn't do one `get` think.

Can we at least make the map allocation when we really need it? Like move that logic to `getCloudDimension` (if possible, would be great to rename it as well)

Reply:
See the last push.  I just got conflated the whole resource attribute -> dimension logic into a single function that is optimized instead of split up.  That is the best way to get performance out of things.

Reply:
Thanks! That looks great